### ISSUE TITLE
Ability to add decorators to the serve method

### ISSUE THREAD

**Issue by m1kola** [wagtail/wagtail#3066]
At the moment, there is no way to add decorators to the `serve` method.

For example, we can't add the `csrf_exempt` decorator. It may be useful for form pages with captcha (because there is no need in csrf token).

**Comment by gasman** [comment]
This was discussed today on the Slack #support channel. It's hard to see how `csrf_exempt` can be made to work properly, as it relies on the CSRF middleware checking the view function for a `csrf_exempt` attribute, and for a Wagtail page request this will always be `wagtail.core.views.serve` - as a result there's no simple way to have it selectively return `csrf_exempt` according to the page it ends up resolving to. Perhaps it would work if the `serve` view was actually a callable object (which _might_ be the same thing as a CBV, but I suspect it isn't...) with `csrf_exempt` defined as an attribute?

As a workaround, CsrfViewMiddleware can be subclassed to add specific page paths as exceptions: https://dpaste.org/Qo6H

**Comment by aleray** [comment]
Dpaste link is exipred. Here is the code:

```python
from django.middleware.csrf import CsrfViewMiddleware

from wagtail.core.views import serve

from my_newsletter.models import NewsletterIndexPage


class CustomCsrfViewMiddleware(CsrfViewMiddleware):

    def process_view(self, request, callback, callback_args, callback_kwargs):

        if callback == serve:
            # We are visiting a wagtail page. Check if this is a NewsletterPage
            # and if so, do not perfom any CSRF validation
            page = NewsletterIndexPage.objects.first()
            path = callback_args[0]

            if page and path.startswith(page.get_url_parts()[-1][1:]):
                return None

        return super().process_view(request, callback, callback_args, callback_kwargs)
```

Register this middleware in place of the CsrfViewMiddleware

```python
MIDDLEWARE = [
    …
    # "django.middleware.csrf.CsrfViewMiddleware",
    "my_newsletter.middleware.CustomCsrfViewMiddleware",
    …
]
```

**Comment by thenewguy** [comment]
Hit this problem as well. Any updates or improvements in the solution?  Would like to able to conditionally enable this based on a modelfield from the page being served

**Comment by thenewguy** [comment]
This could be simplified if the [page fetching logic](https://github.com/wagtail/wagtail/blob/main/wagtail/views.py#L18) was refactored from the serve view into a reusable function that cached the lookup on the request.

Then the middleware could fetch the page if the resolved view function is `wagtail.views.serve`.  Since the page is cached on the request the lookup isn't doubled. If the page meets some criteria middleware can react appropriately.

**Comment by thenewguy** [comment]
Here is a pseudo codish example:

```
from wagtail.views import serve

def find_page_for_request(request, path):
    if getattr(request, '_wagtail_find_page_for_request'):
      # we need a valid Site object corresponding to this request in order to proceed
      site = Site.find_for_request(request)
      if not site:
          raise Http404

      path_components = [component for component in path.split("/") if component]
      page, args, kwargs = site.root_page.localized.specific.route(
           request, path_components
      )
      request._wagtail_find_page_for_request = (page, args, kwargs)
    return request._wagtail_find_page_for_request



def serve(request, path):
    page, args, kwargs = find_page_for_request(request, path)
    
    for fn in hooks.get_hooks("before_serve_page"):
        result = fn(page, request, args, kwargs)
        if isinstance(result, HttpResponse):
            return result

    return page.serve(request, *args, **kwargs)



class SomeWagtailProcessViewMiddleware:
    def process_view(request, view_func, view_args, view_kwargs):
      if serve == view_func:
        page, args, kwargs = find_page_for_request(request, *view_args, **view_kwargs)
        if page meets my criteria:
          view_func.csrf_exempt = True
      return None

```

Thoughts?

**Comment by thenewguy** [comment]
Incase one looks at my example and wonders about adjusting the `view_func.csrf_exempt` back to `False`, here is how you could implement an actual CSRF hook:

```
class CustomCsrfViewMiddleware(CsrfViewMiddleware):
  def process_view(request, view_func, view_args, view_kwargs):
        if serve == view_func:
          page, args, kwargs = find_page_for_request(request, *view_args, **view_kwargs)
          if page meets my criteria:
            self._accept(request) # or just return None
        return super().process_view(request, view_func, view_args, view_kwargs)
```

**Comment by thenewguy** [comment]
@zerolab @gasman This is the issue/solution I was referring to. Do you like it?

**Comment by zerolab** [comment]
Fixed via #11683
